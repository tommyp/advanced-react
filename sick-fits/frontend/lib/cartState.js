/* eslint-disable import/prefer-default-export */
import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  // This is our own custom provider!
  // We will store data (state) and functionality
  // (updaters) in here and anyone can access it via the consumer

  const [cartOpen, setCartOpen] = useState(false);

  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  return (
    <LocalStateProvider value={{ cartOpen, toggleCart }}>
      {children}
    </LocalStateProvider>
  );
}

// make a custom hook for accessing the cart local state

function useCart() {
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, useCart };
