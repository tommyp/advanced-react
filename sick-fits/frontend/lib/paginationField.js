/* eslint-disable no-plusplus */
import { PAGINATION_QUERY } from '../components/Pagination';

export default function paginationField() {
  return {
    keyArgs: false, // tells apollo we will take care of everything
    read(existing = [], { args, cache }) {
      const { skip, first } = args;
      // read the number of items on the page from the cache
      const data = cache.readQuery({ query: PAGINATION_QUERY });

      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      // check if we have existing items
      const items = existing.slice(skip, skip + first).filter((x) => x);
      if (items.length && items.length !== first && page === pages) {
        return items;
      }

      if (items.length !== first) {
        // we don't have any items, we must go to the network
        return false;
      }

      // if there are items, just return them from the cache and we don't need to go to the network

      // if there are items AND
      // there aren't enough items to satisfy how many were request
      // AND if we're on the last page

      if (items.length) {
        return items;
      }

      return false;
      // first thing it does it asks the read function for those items

      // we can either do one of two things

      // 1st: return the items from the cache

      // 2nd: return false from here, which causes a network request
    },
    merge(existing, incoming, { args }) {
      // this runs when the apollo client comes back from the network with our products
      const { skip, first } = args;

      const merged = existing ? existing.slice(0) : [];

      for (let i = skip; i < skip + incoming.length; ++i) {
        merged[i] = incoming[i - skip];
      }

      return merged;
    },
  };
}
