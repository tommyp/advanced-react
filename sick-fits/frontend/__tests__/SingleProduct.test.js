import { MockedProvider } from '@apollo/react-testing';
import { render, screen } from '@testing-library/react';
import SingleProduct, { SINGLE_ITEM_QUERY } from '../components/SingleProduct';
import { fakeItem } from '../lib/testUtils';

const mockProduct = fakeItem();
const mocks = [
  {
    request: { query: SINGLE_ITEM_QUERY, variables: { id: 'abc123' } },
    result: {
      data: {
        product: mockProduct,
      },
    },
  },
  {
    request: { query: SINGLE_ITEM_QUERY, variables: { id: '404' } },
    result: {
      errors: [{ message: 'Item not found!!!' }],
    },
  },
];

describe('<SingleProduct />', () => {
  it('renders correctly', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <SingleProduct id={mockProduct.id} />
      </MockedProvider>,
    );

    await screen.findByTestId('singleProduct');

    expect(container).toMatchSnapshot();
  });

  it('errors out when an item is not found', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <SingleProduct id="404" />
      </MockedProvider>,
    );

    await screen.findByTestId('graphql-error');

    expect(container).toHaveTextContent('Shoot!');
    expect(container).toHaveTextContent('Item not found!!!');
  });
});
