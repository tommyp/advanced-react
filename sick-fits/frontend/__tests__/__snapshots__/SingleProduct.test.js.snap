// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleProduct /> errors out when an item is not found 1`] = `
<div>
  <div
    class="ErrorMessage__ErrorStyles-sc-11u5fgj-0 cEfbva"
  >
    <p
      data-testid="graphql-error"
    >
      <strong>
        Shoot!
      </strong>
      No more mocked responses for the query: query SINGLE_ITEM_QUERY($id: ID!) {
  product: Product(where: {id: $id}) {
    name
    price
    description
    id
    photo {
      image {
        publicUrlTransformed
        __typename
      }
      __typename
    }
    __typename
  }
}

Expected variables: {"id":404}

Failed to match 2 mocks for this query, which had the following variables:
  {"id":"abc123"}
  {"id":"404"}

    </p>
  </div>
</div>
`;

exports[`<SingleProduct /> renders correctly 1`] = `
<div>
  <div
    class="SingleProduct__ProductStyles-sc-1kapz8w-0 jnlTon"
    data-testid="singleProduct"
  >
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <div
      class="details"
    >
      <h2>
        dogs are best
      </h2>
      <p>
        dogs
      </p>
    </div>
  </div>
</div>
`;
